### PR TITLE
3752 validate usernames

### PR DESCRIPTION
**Pull request by ferrys**
This PR adds support for bean validation for usernames and fixes the partial creation of accounts.

## Related Issues

- connects to #3752 : Sign up: Does not validate usernames and partially creates account if invalid

## Pull Request Checklist

- [x] Unit [tests][] completed
- [x] Integration [tests][]: None
- [x] Deployment requirements, [SQL updates][], [Solr updates][], etc.: None
- [ ] [Documentation][docs] completed
- [ ] Merged latest from "develop" [branch][] and resolved conflicts

[tests]: http://guides.dataverse.org/en/latest/developers/testing.html
[SQL updates]: https://github.com/IQSS/dataverse/tree/develop/scripts/database/upgrades
[Solr updates]: https://github.com/IQSS/dataverse/blob/develop/conf/solr/4.6.0/schema.xml
[docs]: http://guides.dataverse.org/en/latest/developers/documentation.html
[branch]: http://guides.dataverse.org/en/latest/developers/branching-strategy.html

### CODE REVIEW

File: src/main/java/edu/harvard/iq/dataverse/authorization/providers/builtin/BuiltinUser.java [hunk 1]
```diff
@@ -43,7 +44,7 @@
 
     @NotBlank(message = "Please enter a username.")
     @Size(min=2, max=60, message ="Username must be between 2 and 60 characters.")
-    @Pattern(regexp = "[a-zA-Z0-9\\_\\-\\.]*", message = "Found an illegal character(s). Valid characters are a-Z, 0-9, '_', '-', and '.'.")
+    @ValidateUserName(message = "Found an illegal character(s). Valid characters are a-Z, 0-9, '_', '-', and '.'.")
```
Comment (bsilverstein95) [inline src/main/java/edu/harvard/iq/dataverse/authorization/providers/builtin/BuiltinUser.java]: Would be good to bundle-ize these messages if possible. Message added at line 186 in `Bundle.properties` is similar, could be neighbors.

File: src/main/java/edu/harvard/iq/dataverse/UserNameValidator.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+     */
+    public static boolean isUserNameValid(final String username, ConstraintValidatorContext context) {
+        //TODO: What other characters do we need to support?
+        String validCharacters = "[a-zA-ZÀ-ÿ0-9\\_\\-\\.";
+        //support accents
+        validCharacters += "\\u00C0-\\u017F";
+        //support chinese characters
+        validCharacters += "\\x{4e00}-\\x{9fa5}";
+        //end
+        validCharacters += "]*";
```
Comment (raprasad) [inline src/main/java/edu/harvard/iq/dataverse/UserNameValidator.java]: Change `*` to `+` at end of line 44.  If not, empty string is valid

File: src/test/java/edu/harvard/iq/dataverse/UserNameValidatorTest.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        assertEquals(false, UserNameValidator.isUserNameValid(" sarah", null));
+        //trailing white space
+        assertEquals(false, UserNameValidator.isUserNameValid("sarah ", null));
+        
+        //symbols!
+        assertEquals(false, UserNameValidator.isUserNameValid("sarah!", null));
+        assertEquals(false, UserNameValidator.isUserNameValid("sarah?", null));
+        assertEquals(false, UserNameValidator.isUserNameValid("sarah:(", null));
+        assertEquals(false, UserNameValidator.isUserNameValid("💲🅰️®️🅰️🚧", null));
+
```
Comment (raprasad) [inline src/test/java/edu/harvard/iq/dataverse/UserNameValidatorTest.java]: Add tests for `null` and empty string `""`

File: src/main/java/edu/harvard/iq/dataverse/UserNameValidator.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        return isUserNameValid(value, context);
+    }
+
+    /**
+     * Here we will validate the username
+     *
+     * @param username
+     * @return boolean 
+     */
+    public static boolean isUserNameValid(final String username, ConstraintValidatorContext context) {
```
Comment (raprasad) [inline src/main/java/edu/harvard/iq/dataverse/UserNameValidator.java]: Add check for null
```java
if (username==null){
  return false;
}

**Comment by raprasad** [comment]
Added a couple of minor corrections

File: src/main/java/edu/harvard/iq/dataverse/authorization/providers/builtin/BuiltinUser.java (hunk 1 above)
Comment (sekmiller) [inline src/main/java/edu/harvard/iq/dataverse/authorization/providers/builtin/BuiltinUser.java]: If you look at the validateUserName method in DataverseUserPage line 220 shows that the error message displayed on the page comes from the Bundle.

**Comment by sekmiller** [comment]
Validator is not checking for username length. If I put in a one character user name, I can get the same result as before - partially create then told username taken.

File: src/main/java/edu/harvard/iq/dataverse/authorization/providers/builtin/BuiltinUser.java (hunk 1 above)
Comment (bsilverstein95) [inline src/main/java/edu/harvard/iq/dataverse/authorization/providers/builtin/BuiltinUser.java]: @sekmiller huh, good catch! Although the message(s) defined above in each tag don't seem to be in the bundle..? `DataverseUserPage.java` only uses "user.username.taken" and "user.username.invalid" -- which are respectively _"This username is already taken."_ and _"This username contains an invalid character."_ -- shouldn't the bundle at least be updated even if the bean validation tags are left with the strings above?

File: src/main/java/edu/harvard/iq/dataverse/UserNameValidator.java (hunk 4 above)
Comment (sekmiller) [inline src/main/java/edu/harvard/iq/dataverse/UserNameValidator.java]: We had been ignoring the length requirement of 2-60 characters, so fixing that will obviate the need for a null check.

Comment (ferrys) [inline src/main/java/edu/harvard/iq/dataverse/UserNameValidator.java]: The matcher throws a null pointer exception if it encounters null, so I added it.
